src/serde.rs

```diff
@@ -1,20 +1,76 @@
 //! Multihash Serde (de)serialization
 
-use std::fmt;
+use core::{fmt, ptr, slice};
 
 use serde::{
     de::{self, SeqAccess, Visitor},
-    Deserialize, Deserializer, Serialize, Serializer,
+    ser, Deserialize, Deserializer, Serialize, Serializer,
 };
 
 use crate::Multihash;
 
+/// The maximum serialization size of `code` is 9 bytes (a large varint encoded u64) and for `size`
+/// is 2 bytes  (a large varint encoded u8), this makes a total of 11 bytes.
+const MAXIMUM_PREFIX_SIZE: usize = 11;
+
+/// The is currently no way to allocate an array that is some constant size bigger then a given
+/// const generic. Once `generic_const_exprs` are a thing, this struct will no longer be needed.
+/// Until then we introduce a hack. We allocate a struct, which contains two independent arrays,
+/// which can be specified with const generics. We then treat the whole struct as a slice of
+/// continuous memory.
+#[repr(C, packed)]
+struct Buffer<const SIZE_FIRST: usize, const SIZE_SECOND: usize> {
+    first: [u8; SIZE_FIRST],
+    second: [u8; SIZE_SECOND],
+}
+
+#[allow(unsafe_code)]
+impl<const SIZE_FIRST: usize, const SIZE_SECOND: usize> Buffer<SIZE_FIRST, SIZE_SECOND> {
+    fn new() -> Self {
+        let buffer = Self {
+            first: [0; SIZE_FIRST],
+            second: [0; SIZE_SECOND],
+        };
+
+        // Make sure that the struct allocated continuous memory, as we exploit that fact with the
+        // `as_slice` and `as_mut_slice()` methods.
+        let start_first = ptr::addr_of!(buffer.first) as *const u8;
+        let start_second = ptr::addr_of!(buffer.second) as *const u8;
+        unsafe {
+            // This should never happen, hence it's OK to panic.
+            if start_second.offset_from(start_first) != SIZE_FIRST as isize {
+                panic!("alignment of the struct is wrong")
+            }
+        }
```

Comment:
This shouldn't be necessary as `packed` guarantees this.

Reply:
But you could consider writing a _test_ to assert this (but I'd avoid doing this at runtime).

Reply:
Actually, you can [`static_assertions`](https://docs.rs/static_assertions/latest/static_assertions) to assert that `mem::size_of::<Self>() == SIZE_FIRST + SIZE_SECOND` at compile time.

Reply:
You can also assert tat `Self` and `[u8]` have the same alignment.

Reply:
I've tried

```rust
const_assert!(mem::size_of::<Self>() == SIZE_FIRST + SIZE_SECOND);
```

but sadly it cannot be compiled:

```console
> cargo build --no-default-features --features serde
error[E0401]: can't use generic parameters from outer function
  --> src/serde.rs:38:38
   |
29 | impl<const SIZE_FIRST: usize, const SIZE_SECOND: usize> Buffer<SIZE_FIRST, SIZE_SECOND> {
   | ---- `Self` type implicitly declared here, by this `impl`
...
38 |         const_assert!(mem::size_of::<Self>() == SIZE_FIRST + SIZE_SECOND);
   |                                      ^^^^
   |                                      |
   |                                      use of generic parameter from outer function
   |                                      use a type here instead

error[E0401]: can't use generic parameters from outer function
  --> src/serde.rs:38:49
   |
29 | impl<const SIZE_FIRST: usize, const SIZE_SECOND: usize> Buffer<SIZE_FIRST, SIZE_SECOND> {
   |            ---------- const parameter from outer function
...
38 |         const_assert!(mem::size_of::<Self>() == SIZE_FIRST + SIZE_SECOND);
   |                                                 ^^^^^^^^^^ use of generic parameter from outer function

error[E0401]: can't use generic parameters from outer function
  --> src/serde.rs:38:62
   |
29 | impl<const SIZE_FIRST: usize, const SIZE_SECOND: usize> Buffer<SIZE_FIRST, SIZE_SECOND> {
   |                                     ----------- const parameter from outer function
...
38 |         const_assert!(mem::size_of::<Self>() == SIZE_FIRST + SIZE_SECOND);
   |                                                              ^^^^^^^^^^^ use of generic parameter from outer function

For more information about this error, try `rustc --explain E0401`.
error: could not compile `multihash` (lib) due to 3 previous errors
```

So would you prefer if I remove that run-time check and write a test instead?

Reply:
I've moved it into a test. Though if anyone has an idea on how to assert it at compile-time, that would even be better.

Reply:
Ah... yeah, generics. I... don't. But I think a test is probably enough.